xdsl/transforms/loop_invariant_code_motion.py

```diff
@@ -0,0 +1,42 @@
+"""
+This pass hoists operation that are invariant to the loops.
+
+Similar to MLIR's loop invariant code motion:
+
+https://mlir.llvm.org/doxygen/LoopInvariantCodeMotion_8cpp_source.html
+
+An operation is loop-invariant if it depends only of values defined outside of the loop.
+LICM moves these operations out of the loop body so that they are not computed more than
+once.
+
+  for i in range(x, N, M):                for i in range(x, N, M):
+    for j in range(0, M, K):    ---->        c[i]= A[1] + b[1]
+      c[i]=A[1]+b[1]
+"""
+
+from xdsl.ir import Operation, Region
+from xdsl.traits import IsTerminator
+
+
+def can_be_hoisted(op: Operation, target_region: Region) -> bool | None:
+    """
+    Checks whether the given op can be hoisted by checking that
+    - the op and none of its contained operations depend on values inside of the
+     loop.
+    """
+    #  Do not move terminators.
+    if op.has_trait(IsTerminator):
+        return False
+
+    # Walk the nested operations and check that all used values are either
+    # defined outside of the loop or in a nested region, but not at the level of
+    # the loop body.
+    for child in op.walk():
+        for operand in child.operands:
+            operand_owner = operand.owner
+            assert operand_owner is not None
+            if op.is_ancestor(operand_owner):
+                continue
```

Comment:
I just ran the test locally, and this line is never hit.

Reply:
I pushed my suggested changes to `sasha/licm/can-be-hoisted`. I added breakpoints to every line in this function, and added test cases that would hit every one of them. I also removed scf and func dialects from the test, since they're not used by this helper. I would recommend reviewing those changes, and merging them to your branch if you're happy with them.

Reply:
Then we can move onto the next step

Reply:
I reused your testcases, Thank you Sasha. Can I move to add IsMemoryEffectFree helper function?